datastore/longhorn.go

```diff
@@ -1301,6 +1318,158 @@ func GetOwnerReferencesForBackingImageManager(backingImageManager *longhorn.Back
+// verifies update
+func (s *DataStore) UpdateBackingImageDataSourceStatus(backingImageDataSource *longhorn.BackingImageDataSource) (*longhorn.BackingImageDataSource, error) {
+	obj, err := s.lhClient.LonghornV1beta1().BackingImageDataSources(s.namespace).UpdateStatus(backingImageDataSource)
+	if err != nil {
+		return nil, err
+	}
+	verifyUpdate(backingImageDataSource.Name, obj, func(name string) (runtime.Object, error) {
+		return s.getBackingImageDataSourceRO(name)
+	})
+	return obj, nil
+}
+
+// DeleteBackingImageDataSource won't result in immediately deletion since finalizer was
+// set by default
+func (s *DataStore) DeleteBackingImageDataSource(name string) error {
+	propagation := metav1.DeletePropagationForeground
+	return s.lhClient.LonghornV1beta1().BackingImageDataSources(s.namespace).Delete(name, &metav1.DeleteOptions{PropagationPolicy: &propagation})
+}
+
+// RemoveFinalizerForBackingImageDataSource will result in deletion if DeletionTimestamp was set
+func (s *DataStore) RemoveFinalizerForBackingImageDataSource(obj *longhorn.BackingImageDataSource) error {
+	if !util.FinalizerExists(longhornFinalizerKey, obj) {
+		// finalizer already removed
+		return nil
+	}
+	if err := util.RemoveFinalizer(longhornFinalizerKey, obj); err != nil {
+		return err
+	}
+	_, err := s.lhClient.LonghornV1beta1().BackingImageDataSources(s.namespace).Update(obj)
+	if err != nil {
+		// workaround `StorageError: invalid object, Code: 4` due to empty object
+		if obj.DeletionTimestamp != nil {
+			return nil
+		}
+		return errors.Wrapf(err, "unable to remove finalizer for backing image data source %v", obj.Name)
+	}
+	return nil
+}
+
+func (s *DataStore) getBackingImageDataSourceRO(name string) (*longhorn.BackingImageDataSource, error) {
+	return s.bidsLister.BackingImageDataSources(s.namespace).Get(name)
+}
+
+// GetBackingImageDataSource returns a new BackingImageDataSource object for the given name and
+// namespace
+func (s *DataStore) GetBackingImageDataSource(name string) (*longhorn.BackingImageDataSource, error) {
+	resultRO, err := s.getBackingImageDataSourceRO(name)
+	if err != nil {
+		return nil, err
+	}
+	// Cannot use cached object from lister
+	return resultRO.DeepCopy(), nil
+}
+
+// ListBackingImageDataSources returns object includes all BackingImageDataSource in namespace
+func (s *DataStore) ListBackingImageDataSources() (map[string]*longhorn.BackingImageDataSource, error) {
+	itemMap := map[string]*longhorn.BackingImageDataSource{}
+
+	list, err := s.bidsLister.BackingImageDataSources(s.namespace).List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	for _, itemRO := range list {
+		// Cannot use cached object from lister
+		itemMap[itemRO.Name] = itemRO.DeepCopy()
+	}
+	return itemMap, nil
+}
+
+// ListBackingImageDataSourcesByNode returns object includes all BackingImageDataSource in namespace
+func (s *DataStore) ListBackingImageDataSourcesByNode(nodeName string) (map[string]*longhorn.BackingImageDataSource, error) {
```

Comment:
NIT: I like to add a generic list function that gets passed a selector, see the refactor I done for the `listBackingImageManagers(selector labels.Selector)` as an example. This way both the list Everything an ByNode both go through the same retrieval + copy code flow.